swagger.yaml

```diff
@@ -1635,7 +1635,7 @@ paths:
           name: file_name
           description: The file to be downloaded.
           type: string
-          enum: [discovery.ign, ipxe-script]
+          enum: [discovery.ign, ipxe-script, static-network-config]
```

Comment:
This endpoint is becoming a bit crowded. At what point should we consider something not a "file" and give it a different endpoint under `/infra-envs/{infra_env_id}/downloads/`

Already these three cases are implemented completely differently. Curious what @avishayt thinks.

Reply:
What about having one common endpoint such as `/v2/infra-envs/{infra_env_id}/downloads` with a query for the item you'd like to download? E.g. `/v2/infra-envs/{infra_env_id}/downloads?artifact=minimal-initrd` or `/v2/infra-envs/{infra_env_id}/downloads?artifact=ipxe-script`

Reply:
>What about having one common endpoint

This is effectively what we have now, right? Are there any other sub-paths under `/downloads`?

> Should we allow users to download static config files by host?

We apply the same static network config to all hosts since this is the discovery image.

Reply:
> > What about having one common endpoint
> 
> This is effectively what we have now, right? Are there any other sub-paths under `/downloads`?

`/downloads/minimal-initrd` is a different subpath. I meant combining that and the `/downloads/files` path 
> > Should we allow users to download static config files by host?
> 
> We apply the same static network config to all hosts since this is the discovery image.
Oh gotcha, then a follow up question: is it useful to generate and send all the network config files as opposed to just the one that's applied to all hosts?

Reply:
> I meant combining that and the /downloads/files path

This is what I'm considering a problem. We're downloading too many unrelated things each with their own separate options from this one endpoint. I'm thinking each should have a separate endpoint, but we can't do anything about the ones that are already here.

I'm not sure if starting with a new endpoint for this will make thinks better or worse :worried:

Reply:
>is it useful to generate and send all the network config files

What other ones? The scripts and such? I think the use for this endpoint is mostly just so admins can verify that the network manager config files look as they would expect.

Reply:
> This is what I'm considering a problem. We're downloading too many unrelated things each with their own separate options from this one endpoint. I'm thinking each should have a separate endpoint, but we can't do anything about the ones that are already here.
> 
> I'm not sure if starting with a new endpoint for this will make thinks better or worse 😟

Oh I see, is it usually better to have different endpoints for each type of download? So it would be split into
`/downloads/ipxe-script`
`/downloads/minimal-initrd`
`/downloads/discovery.ign`

Would this API definition benefit the end-user, be easier to maintain, etc?

Reply:
I think the problem right now is that some url params on this point only apply when you're downloading certain files and that's just not a great experience. @avishayt brought this up when we added the boot control stuff.

That said, I don't think we can change things like the ipxe script and discovery ignition without breaking the current API so I'm not sure how much we gain by splitting just this one out :man_shrugging:

Reply:
Would it be a good idea to start separating the APIs now instead of adding more to this `/downloads/files` endpoint?

---
What would be the process if we were to create new separate API endpoints for discovery ignition and ipxe script?

Would it be to introduce these APIs and keep (but start deprecating) the original downloads/files API for them?

Reply:
In the cluster download API it made sense to have a single endpoint.  We have 10 files there available for download so I wanted to avoid 10 download APIs.  The implementation shouldn't affect the API, but all those files are stored in S3 so the implementation for all 10 is identical - just pass the right file name.  We split out 'credentials' because that API has different authorization than 'files'.
Here I understand that the implementation is different but that shouldn't affect the API.  It would be better to keep things under 'files' to be consistent with cluster downloads.  In that case, we should move minimal-initrd to 'files' and deprecate the current API.  I'm also OK with making separate endpoints if you feel strongly about it, but then we should deprecate 'files'.

Reply:
:+1: lets just keep it the same as clusters then.

Easier to deprecate one endpoint and move it here than create a bunch of others.

I remember having a conversation with someone, I thought it was @avishayt, about how the parameters only apply if other parameters are set in a certain way, but I agree that's it's not worth the trouble.